app/src/main/java/com/gh4a/adapter/BookmarkAdapter.java

```diff
@@ -105,31 +133,52 @@ public int getItemCount() {
 
     @Override
     public long getItemId(int position) {
-        return mCursor.moveToPosition(position) ? mCursor.getLong(mIdColumnIndex) : -1;
+        return moveCursorToPosition(position) ? mCursor.getLong(mIdColumnIndex) : -1;
     }
 
     @Override
     public void onClick(View view) {
         ViewHolder vh = (ViewHolder) view.getTag();
         int position = vh.getAdapterPosition();
-        if (position != RecyclerView.NO_POSITION && mCursor.moveToPosition(position)) {
-            mItemClickListener.onItemClick(mCursor.getLong(mIdColumnIndex),
+        if (position != RecyclerView.NO_POSITION && moveCursorToPosition(position)) {
+            mItemInteractListener.onItemClick(mCursor.getLong(mIdColumnIndex),
                     mCursor.getString(mUrlColumnIndex));
         }
     }
 
-    @Override
-    public boolean onLongClick(View view) {
-        ViewHolder vh = (ViewHolder) view.getTag();
-        int position = vh.getAdapterPosition();
-        if (position != RecyclerView.NO_POSITION && mCursor.moveToPosition(position)) {
-            mItemClickListener.onItemLongClick(mCursor.getLong(mIdColumnIndex));
-            return true;
+    public void onItemMoved(int fromPos, int toPos) {
+        mPositions.add(toPos, mPositions.remove(fromPos));
+        notifyItemMoved(fromPos, toPos);
+    }
+
+    public void onItemSwiped(RecyclerView.ViewHolder viewHolder) {
+        int position = viewHolder.getAdapterPosition();
+        if (position == RecyclerView.NO_POSITION || !moveCursorToPosition(position)) {
+            return;
         }
-        return false;
+
+        final Context context = viewHolder.itemView.getContext();
+        final long id = mCursor.getLong(mIdColumnIndex);
+        final String name = mCursor.getString(mNameColumnIndex);
+        final int type = mCursor.getInt(mTypeColumnIndex);
+        final String url = mCursor.getString(mUrlColumnIndex);
+        final String extraData = mCursor.getString(mExtraColumnIndex);
+
+        updateOrder(context);
+
+        Uri uri = ContentUris.withAppendedId(BookmarksProvider.Columns.CONTENT_URI, id);
+        context.getContentResolver().delete(uri, null, null);
+
+        Snackbar.make(viewHolder.itemView, R.string.bookmark_removed, Snackbar.LENGTH_LONG)
+                .setAction(R.string.undo, new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        BookmarksProvider.saveBookmark(context, name, type, url, extraData, false);
+                    }
+                }).show();
     }
 
-    static class ViewHolder extends RecyclerView.ViewHolder {
+    class ViewHolder extends RecyclerView.ViewHolder implements View.OnTouchListener {
```

Comment:
I think I'd prefer passing the interact listener into the ViewHolder's constructor over making this non-static.